Workspaces: diff.patch view (#2314)
wandb/docs: models/app/features/panels/code.mdx

```python
import wandb

with wandb.init() as run:
    run.log_code(
         root="../",
         include_fn=lambda path: path.endswith(".py") or path.endswith(".ipynb"),
         exclude_fn=lambda path, root: os.path.relpath(path, root).startswith(
             "cache/"
         ),
    )
```

For more control over the types and locations of source code files that are saved, see the [reference docs](/models/ref/python/experiments/run#log_code).

### Automatically capture code

Use `wandb.init(settings=wandb.Settings(code_dir="."))` to automatically capture all code in the current directory and subdirectories. By default, W&B captures:

- Files ending in `.py`
- requirements.txt
- Dockerfile
- Excludes anything under `wandb/` or `.wandb/`

The following example shows how capture all code in the current directory and subdirectories:

```python
import wandb

with wandb.init(settings=wandb.Settings(code_dir=".")) as run:
    # Your training code here
```

You can also specify a different directory by replacing `.` with the path to the directory that you want to capture.

## View local changes

> View local changes in your project workspace. W&B saves the diff of unstaged changes across runs in a `diff.patch` or `diff_<sha>.patch` file in project workspace.

Reviewer: It's uncommitted changes, not unstaged changes. Thread here: https://github.com/wandb/docs/pull/2314#discussion_r3185578108

> captures all uncommitted changes — both staged and unstaged — relative to HEAD 
> 
> There's also a second patch file generated: diff_{sha}.patch (handler.go:622-628), which runs git diff @{u}         
>   --submodule=diff — that one diffs against the upstream tracking branch, capturing unpushed commits + uncommitted    
>   changes.

Author: @julia-g-rose Confused here, from the that same thread I was under the assumption you wanted it called "Local changes"?

Reviewer: You said "W&B saves the diff of unstaged change". This isn't correct, it's both staged and unstaged

Reviewer: captures all uncommitted changes — both staged and unstaged — relative to HEAD

> W&B generates diff files when you log code with W&B and have unstaged changes in your git repository. To view the diff of unstaged changes across runs:

Reviewer: uncommitted, not unstaged

Reviewer: What I mean here is just that saying "unstaged changes" implies we only capture unstaged changes, which isn't true

1. Navigate to your project's workspace.
2. Click on the **Workspace** tab in the left sidebar. 
3. From the run list, select the run that you want to view the diff of unstaged changes for. This will open the run overview page.
2. Select the **Files** tab.
3. From the list of files, click on the `diff.patch` or the `diff_<sha>.patch` file.

> 4. Select a file from the list to expand the diff of unstaged changes for that file.

Reviewer: uncommitted, not unstaged

keywords: ["log_code", "code_dir", "code comparer", "Jupyter notebook history", "reproducibility"]
---

{/* By default, W&B only saves the latest git commit hash. You can turn on more code features to compare the code between your experiments dynamically in the UI. */}

Use W&B to save your code, compare code across runs, view local changes, and capture Jupyter session history.

## Enable code saving

Configure code saving for your team or organization. Team and organization controls are documented in [Configure privacy settings](/platform/hosting/privacy-settings).

<Info>

> By default, W&B disables code saving for all teams. Before you can turn it on for a team, an organization admin must turn it on for the organization. See the [Organization](#organization) section.

Reviewer 2: ```suggestion
By default, W&B disables code saving for all teams. Before you can turn it on for a team, an organization admin must enable it at the organization level. See the [Organization](#organization) section.
```
Reducing word echo

</Info>

### Team

> Team admins open the team **Settings** page, go to the **Privacy** section, and configure **Enable code saving by default** for runs in that team. This option is available only when an organization admin has not enforced code saving restrictions for the whole organization. For navigation steps, see [Configure privacy settings for a team](/platform/hosting/privacy-settings#configure-privacy-settings-for-a-team).

Reviewer 2: ```suggestion
To enable code saving for a team as a team admin, navigate to the **Settings** page, then go to the **Privacy** section, and configure **Enable code saving by default** for runs in that team. This option is available only when an organization admin has not enforced code saving restrictions for the whole organization. For navigation steps, see [Configure privacy settings for a team](/platform/hosting/privacy-settings#configure-privacy-settings-for-a-team).
```

### Organization

> Organization admins open organization **Settings**, go to the **Privacy** section, and can turn on **Enforce default code saving restrictions** so code saving stays off by default for every team. While this enforcement is on, team admins cannot turn on **Enable code saving by default** for a team. For the full list of organization controls, see [Enforce privacy settings for all teams](/platform/hosting/privacy-settings#enforce-privacy-settings-for-all-teams).

Reviewer 2: ```suggestion
To enable code saving for an organization as an organization admin, navigate to **Settings**, go to the **Privacy** section, and turn on **Enforce default code saving restrictions** so code saving stays off by default for every team. While this enforcement is on, team admins cannot turn on **Enable code saving by default** for a team. For the full list of organization controls, see [Enforce privacy settings for all teams](/platform/hosting/privacy-settings#enforce-privacy-settings-for-all-teams).
```

## Save code

{/* When you enable code saving, W&B saves the code from the file that called `wandb.init()`. To save additional library code, you have three options: */}

Capture source code that produced a run as an [artifact](/models/ref/python/experiments/artifact). The code artifact is viewable within your project's workspace.

> There are two ways approaches to save code: a fine-grained approached and an automatic approach.

Reviewer 2: ```suggestion
There are two ways to save code: a fine-grained approach and an automatic approach.
```

### Fine grain control

Use `wandb.Run.log_code()` to log specific files or directories as an [artifact](/models/ref/python/experiments/artifact). 

By default, W&B walks through the current directory and logs all files that end with `.py`.

The following example shows how to use `wandb.Run.log_code()` to log the current directory:

```python
import wandb

with wandb.init() as run:
    # Log the current directory as a code artifact
    run.log_code(root=".")
```

The following example shows how to use `wandb.Run.log_code()` with `include_fn=` and `exclude_fn=` parameters to specify which files to include and exclude when you log code:

```python
import wandb

with wandb.init() as run:
    run.log_code(
         root="../",
         include_fn=lambda path: path.endswith(".py") or path.endswith(".ipynb"),
         exclude_fn=lambda path, root: os.path.relpath(path, root).startswith(
             "cache/"
         ),
    )
```

> For more control over the types and locations of source code files that are saved, see the [reference docs](/models/ref/python/experiments/run#log_code).

Reviewer 2: ```suggestion
For more control over the types and locations of source code files that W&B saves, see the [reference docs](/models/ref/python/experiments/run#log_code).
```
Makes the sentence a little more active.

### Automatically capture code

Use `wandb.init(settings=wandb.Settings(code_dir=))` to automatically capture all code in the current directory and subdirectories. By default, W&B captures:

- Files ending in `.py`
- requirements.txt
- Dockerfile
- Excludes anything under `wandb/` or `.wandb/`

The following example shows how capture all code in the current directory and subdirectories:

```python
import wandb

with wandb.init(settings=wandb.Settings(code_dir=".")) as run:
    # Your training code here
```

You can also specify a different directory by replacing `.` with the path to the directory that you want to capture.

## View local changes

> View local changes in your project workspace. W&B generates diff files when you log code with staged or unstaged changes in your Git repository.

Reviewer 2: The intro line kinda needlessly reiterates the title. I think I would go with something like this:

```suggestion
W&B generates diff files when you log code with staged or unstaged changes in your Git repository, and you can view the changes in the W&B UI.
```
This is not perfect, but I'd opt for something that intros the section better.

W&B names the diff files `diff.patch` or `diff_<sha>.patch`. These files contain local code changes that are not yet committed to Git, relative to `HEAD`.

To view diff files:

> 1. Navigate to your project's workspace.

Reviewer 2: These all look misnumbered.

    <img src="/images/app_ui/code_comparer.png" alt="Code comparer panel"  />
</Frame>

> ## View unstaged changes

Reviewer: Can we also mention that in addition to just viewing the uncommitted changes, you can download the diff.patch file and apply the patch to your repo? This is a common workflow. There is also the other option to "Copy file contents", which can be helpful for applying a patch to a remote machine

> View unstaged changes in your project workspace.

Reviewer: @nhusher just to clarify terminology, do we capture unstaged changes in the diff patch or uncommitted changes? I thought we capture uncommitted changes, which is broader than unstaged changes

Reviewer 3: We capture unstaged changes, not uncommitted changes.

Reviewer: Just want to double-check? Only asking because I saw a different response from Claude, sharing more details here:

Claude says we are picking up both:

<img width="698" height="242" alt="Screenshot 2026-05-04 at 6 36 06 PM" src="https://github.com/user-attachments/assets/9e457c68-de5e-4f7a-bf73-cc47ced6b98e" />

And Claude Code also found notes in the repo stating that it's "uncommitted" changes:
<img width="674" height="294" alt="Screenshot 2026-05-04 at 6 39 19 PM" src="https://github.com/user-attachments/assets/3ee23542-af5f-42fd-83ab-7aec339e4965" />

More investigation:
 ```
There it is. fileInfoFromName.ts:28-29 has the answer in the comment:                                               
                                                                                                                      
  ▎ diff.patch and diff_{sha}.patch are diffs generated by the SDK when there are unstaged or unpushed changes in the 
  ▎ git repo when the run was started.
                                                                                                                      
  So it captures both unstaged changes and unpushed commits — not just uncommitted edits, but also commits that       
  haven't been pushed to the remote yet. The actual generation happens in the W&B SDK (not in this repo).  
```

Final investigation:
```
Found it. Here's the definitive answer:                                                                             
                                                                                                                      
  diff.patch is generated by running git diff HEAD --submodule=diff (line 76 in wandb/core/internal/gitops/git.go).   
  This captures all uncommitted changes — both staged and unstaged — relative to HEAD.                                
                                                                                                                      
  There's also a second patch file generated: diff_{sha}.patch (handler.go:622-628), which runs git diff @{u}         
  --submodule=diff — that one diffs against the upstream tracking branch, capturing unpushed commits + uncommitted    
  changes.                                                                                                            
                                                                                                                      
  So the comment in your repo (fileInfoFromName.ts:28-29) saying "unstaged or unpushed" is slightly imprecise:        
  - diff.patch = uncommitted changes (staged + unstaged) vs HEAD
  - diff_{sha}.patch = everything the remote doesn't have (unpushed commits + uncommitted changes) vs upstream  
 ```

Obviously AI can be wrong (especially when prompted by someone who doesn't know the code base), so defer to you for the final verdict

Reviewer 3: That's correct. What it doesn't capture is _untracked_ files. Sorry about that, I was confused yesterday

Reviewer: Thanks for double-checking! Claude recommended calling this "Local changes", cc @ngrayluna 

```
❯ What's the best thing to call what we do? "Capturing uncommitted changes" or a different title?                     
                                                                                                  
⏺ "Uncommitted changes" is accurate for diff.patch, but it doesn't cover diff_{sha}.patch which also includes unpushed  commits. If you want a single label that covers both files, something like "unsynchronized changes" or "local changes" could work — though those are vaguer.

  If you're labeling them separately, I'd go with:                                                                    
  - diff.patch → "uncommitted changes"
  - diff_{sha}.patch → "unpushed changes"                                                                             
                                         
  What's the context — are you updating the UI copy or the code comment?                                              
                                                                                                                      
✻ Worked for 5s                                                                                                       
                                                                                                                      
❯ Product messaging                                                                                                   
                    
⏺ For product messaging, "local changes" is probably the best fit — it's plain language, non-technical, and covers both files without requiring users to understand the git staging model. "Uncommitted changes" and "unpushed changes" are accurate but assume git fluency that not all users will have.
```

Reviewer: Just to double-check, @nhusher the pretty view also applies to diff_{sha}.patch?

Author: TL;DR `diff.*` file is basically Linux's `diff` saved and used by Git.

1. Navigate to your project's workspace.

> 2. Select the **Files** tab.

Reviewer: I'm not sure if we need to mention it, but another entry point is the "View File" button on the "Overview" tab
